storage/zynq7000-flash/flashdrv.c

```diff
@@ -0,0 +1,502 @@
+	/* Size of the data which is received during cmd transfer */
+	dataSz = cmds.dummyCyc / 8;
+
+	memset(fdrv_common.cmdTx, 0, cmdSz);
+	fdrv_common.cmdTx[0] = cmds.opCode;
+
+	qspi_start();
+	if ((res = qspi_transfer(fdrv_common.cmdTx, fdrv_common.cmdRx, cmdSz, TIMEOUT_CMD_MS)) < 0) {
+		qspi_stop();
+		return res;
+	}
+
+	/* Copy data received after dummy bytes */
+	memcpy(cfi, (fdrv_common.cmdRx + cmds.size), dataSz);
+	res = qspi_transfer(NULL, (uint8_t *)cfi + dataSz, sizeof(flash_cfi_t) - dataSz, 5 * TIMEOUT_CMD_MS);
+	qspi_stop();
+
+	return res;
+}
+
+
+static int flashdrv_statusRegGet(unsigned int *val)
+{
+	ssize_t res;
+	size_t cmdSz;
+	const flash_cmd_t *cmd = &fdrv_common.info.cmds[flash_cmd_rdsr1];
+
+	*val = 0;
+
+	cmdSz = cmd->size + cmd->dummyCyc / 8;
+	memset(fdrv_common.cmdTx, 0, cmdSz);
+
+	fdrv_common.cmdTx[0] = cmd->opCode;
+
+	qspi_start();
+	res = qspi_transfer(fdrv_common.cmdTx, (uint8_t *)val, cmdSz, TIMEOUT_CMD_MS);
+	qspi_stop();
+
+	return res;
+}
+
+
+static int flashdrv_wipCheck(time_t timeout)
+{
+	int res;
+	unsigned int st;
+	time_t start = flashdrv_timeMsGet();
+
+	do {
+		if ((res = flashdrv_statusRegGet(&st)) < 0)
+			return res;
+
+		if ((flashdrv_timeMsGet() - start) >= timeout)
+			return -ETIME;
+
+	} while ((st >> 8) & 0x1);
+
+	return EOK;
+}
+
+
+static int flashdrv_welSet(unsigned int cmdID)
+{
+	ssize_t res;
+	const flash_cmd_t *cmd;
+
+	if (cmdID != flash_cmd_wrdi && cmdID != flash_cmd_wren)
+		return -EINVAL;
+
+	cmd = &fdrv_common.info.cmds[cmdID];
+
+	memset(fdrv_common.cmdTx, 0, cmd->size);
+	fdrv_common.cmdTx[0] = cmd->opCode;
+
+	qspi_start();
+	res = qspi_transfer(fdrv_common.cmdTx, fdrv_common.cmdRx, cmd->size, TIMEOUT_CMD_MS);
+	qspi_stop();
+
+	return res;
+}
+
+
+static ssize_t flashdrv_pageProgram(addr_t offs, const void *buff, size_t len)
+{
+	ssize_t res;
+	time_t timeout;
+	const flash_cfi_t *cfi = &fdrv_common.info.cfi;
+	const flash_cmd_t *cmd = &fdrv_common.info.cmds[flash_cmd_pp];
+
+	if (len == 0)
+		return 0;
+
+	if (buff == NULL || len % CFI_SIZE_PAGE(cfi->pageSize))
+		return -EINVAL;
+
+	if ((res = flashdrv_welSet(flash_cmd_wren)) < 0)
+		return res;
+
+	memset(fdrv_common.cmdTx, 0, cmd->size);
+	fdrv_common.cmdTx[0] = cmd->opCode;
+	fdrv_common.cmdTx[1] = (offs >> 16) & 0xff;
+	fdrv_common.cmdTx[2] = (offs >> 8) & 0xff;
+	fdrv_common.cmdTx[3] = offs & 0xff;
+
+	qspi_start();
+	if ((res = qspi_transfer(fdrv_common.cmdTx, fdrv_common.cmdRx, cmd->size, TIMEOUT_CMD_MS)) < 0) {
+		qspi_stop();
+		return res;
+	}
+
+	/* Transmit data to write */
+	if ((res = qspi_transfer(buff, NULL, len, TIMEOUT_CMD_MS)) < 0) {
+		qspi_stop();
+		return res;
+	}
+	qspi_stop();
+
+	len = res;
+
+	timeout = CFI_TIMEOUT_MAX_PROGRAM(cfi->timeoutTypical.pageWrite, cfi->timeoutMax.pageWrite) + TIMEOUT_CMD_MS;
+	if ((res = flashdrv_wipCheck(timeout)) < 0)
+		return res;
+
+	return len;
+}
+
+
+/* Device interface */
+
+void flashdrv_flashInfo(flash_info_t *info)
```

Comment:
Hmmm, maybe 
```suggestion
const flash_info_t* flashdrv_flashInfo(void)
```
instead? We shouldn't really need a copy